### PR TITLE
Sumcheck

### PR DESCRIPTION
**Pull request by mpenciak**
* Adds the Sumcheck verifier protocol
* Adds supporting polynomial libraries
* Adds the sumcheck data contract generated by a python script
* Adds unit tests for the polynomial libraries
* Adds end to end tests for sumcheck in the initial part of the Spartan verifier

Note: The data contract has data for all of the outer, inner, and batched sumchecks, though only the outer data is used in the primary and secondary sumcheck tests. Once computational commitments are implemented, we can write end to end tests that will use the inner and batched data.

### CODE REVIEW

**Comment by storojs72** [comment]
As far as I see, new file `src/sumcheck.sol` is being added and also there are some changes on KeccakTranscript. How do you debug and test your implementation? And what part of Nova sumcheck protocol is being implemented? Just for pointer - we expect [this](https://github.com/lurk-lab/Nova/blob/solidity-verifier/src/spartan/sumcheck.rs#L22) one.

**Comment by storojs72** [comment]
@mpenciak , I added basic unit-testing for primary part of sumcheck computation. There were some issues with using the right modulus and negation. Having this as an example, it should be easier to implement secondary part and update data contract as well as Python generator script.

**Review by huitseeker — APPROVED** [review]
Thanks a lot @mpenciak !

File: src/verifier/step4/SumcheckLogic.sol [hunk 1]
```diff
@@ -0,0 +1,488 @@
...
+        uint256 e = claim;
+        uint256[] memory r = new uint256[](num_rounds);
+
+        uint8[] memory p_label = new uint8[](1);
+        uint8[] memory c_label = new uint8[](1);
+
+        p_label[0] = 112;
+        c_label[0] = 99;
+
+        require(proof.compressed_polys.length == num_rounds, "Wrong number of polynomials");
```
Comment (storojs72) [inline src/verifier/step4/SumcheckLogic.sol:415]: It is very helpful while debugging if we put library and function name into exception string, like following:
`"[PrimarySumcheck::verify] Wrong number of polynomials"`

File: src/verifier/step4/SumcheckLogic.sol [hunk 2]
```diff
@@ -0,0 +1,488 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity ^0.8.0;
+
+import "src/pasta/Pallas.sol";
+import "src/pasta/Vesta.sol";
+import "src/verifier/step4/KeccakTranscript.sol";
+import "src/verifier/step4/EqPolynomial.sol";
+
+library PallasPolyLib {
+    struct MLPoly {
```
Comment (storojs72) [inline src/verifier/step4/SumcheckLogic.sol:10]: As far as I see, these structs (MLPoly, SparseEntry, SparsePoly, UniPoly, CompressedUniPoly) are identical for both PallasPolyLib and VestaPolyLib. Can't we have them defined in some single place?

File: src/verifier/step4/sumcheck-data-contract-gen.py [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+data_contract_body += secondary_outer_data_def
+data_contract_body += secondary_inner_data_def
+data_contract_body += secondary_batch_data_def
+data_contract_body += primary_transcript_data_def
+data_contract_body += secondary_transcript_data_def
+
+data_contract_body += "}\n"
+
+print(header)
+print(data_contract_body)
```
Comment (storojs72) [inline src/verifier/step4/sumcheck-data-contract-gen.py:159]: newline?

File: src/verifier/step4/SumcheckLogic.sol [hunk 4]
```diff
@@ -0,0 +1,488 @@
...
+
+        uint256 result;
+
+        for (uint256 j = 0; j < r.length; j++) {
+            if (a[j]) {
+                // result *= r[j];
+                result = mulmod(result, r[j], Pallas.P_MOD);
+            } else {
+                // result *= (1 - r[j]);
+                result = mulmod(result, addmod(1, Pallas.negateBase(r[j]), Pallas.P_MOD), Pallas.P_MOD);
```
Comment (storojs72) [inline src/verifier/step4/SumcheckLogic.sol:75]: We can make this function generic by passing the correct modulus and negation function as parameters

File: src/verifier/step4/SumcheckLogic.sol [hunk 5]
```diff
@@ -0,0 +1,488 @@
...
+            uint256 a = P.Z[i];
+            uint256 b = P.Z[mid + i];
+            // new_Z[i] = a + r * (b - a);
+            new_Z[i] = addmod(a, mulmod(r, addmod(b, Pallas.negateBase(a), Pallas.P_MOD), Pallas.P_MOD), Pallas.P_MOD);
+        }
+
+        return MLPoly(P.numvars - 1, new_Z);
+    }
+
+    function evaluate(MLPoly calldata P, uint256[] calldata r) public pure returns (uint256 result) {
```
Comment (storojs72) [inline src/verifier/step4/SumcheckLogic.sol:53]: We can make this function generic, by forcing caller to pass correct modulus as parameter

File: src/verifier/step4/SumcheckLogic.sol [hunk 6]
```diff
@@ -0,0 +1,488 @@
...
+
+    struct UniPoly {
+        uint256[] coeffs;
+    }
+
+    struct CompressedUniPoly {
+        uint256[] coeffs_except_linear_term;
+    }
+
+    function newML(uint256[] memory Z, uint64 numvars) public pure returns (MLPoly memory) {
```
Comment (storojs72) [inline src/verifier/step4/SumcheckLogic.sol:33]: I think we don't need two identical `newML` functions in PallasPolyLib and VestaPolyLib as they don't need to know the curve

File: src/verifier/step4/SumcheckLogic.sol [hunk 7]
```diff
@@ -0,0 +1,488 @@
...
+        require(Z.length == 2 ^ numvars);
+        return MLPoly(numvars, Z);
+    }
+
+    function boundPolyVarTop(MLPoly calldata P, uint256 r) public pure returns (MLPoly memory) {
```
Comment (storojs72) [inline src/verifier/step4/SumcheckLogic.sol:38]: Same. By passing modulus as parameter, we can make this function generic

File: src/verifier/step4/SumcheckLogic.sol [hunk 8]
```diff
@@ -0,0 +1,488 @@
...
+            // result += power * poly.coeffs[i];
+            result = addmod(result, mulmod(power, poly.coeffs[i], Pallas.R_MOD), Pallas.R_MOD);
+            // power *= r;
+            power = mulmod(power, r, Pallas.R_MOD);
+        }
+
+        return result;
+    }
+
+    function compress(UniPoly memory poly) public pure returns (CompressedUniPoly memory result) {
```
Comment (storojs72) [inline src/verifier/step4/SumcheckLogic.sol:134]: Do we need compression somewhere? I didn't find its usage. If it is still necessary, we need to test it either explicitly by introducing unit-test or implicitly by using it in some higher level function that is tested.

**Review by storojs72 — CHANGES_REQUESTED** [review]
I like the idea of having bunch of `absorb` functions, each taking its own type of input. It seems to be a bit more ergonomic than what I used in [pp-spartan](https://github.com/lurk-lab/solidity-verifier/blob/pp-spartan/src/NovaVerifierAbstractions.sol#L119) branch, where I have similar bunch of `toTranscriptBytes` functions and I have to introduce a redundant intermediate `uint8[] memory` variable all the time, when I need to invoke `absorb`.

At the same time, after taking draft code from early version of this PR and adapting it for usage in [pp-spartan branch](https://github.com/lurk-lab/solidity-verifier/tree/pp-spartan), I have couple of remarks:

I think a lot of Sumcheck's logic can be implemented in a more generic way: in some places we can pass modulus as a uint256 variable (either Vesta or Pallas), also we can use passing function as a parameter if we need to use particular negation. Also, structures that operates with just `uint256` can be defined only once.

It looks like you implemented some more types of evaluations, however, from my experience, the pp-spartan verification uses only sumcheck over `UniPoly` polynomials ([this](https://github.com/lurk-lab/solidity-verifier/blob/pp-spartan/src/blocks/Sumcheck.sol) implementation is sufficient for e2e pp-spartan verification). Though I didn't check if other polynomials are used while [IPA](https://github.com/lurk-lab/Nova/blob/solidity-verifier-pp-spartan/src/spartan/ppsnark.rs#L2031) (as it is going to be replaced with KZG/Zeromorph eventually). Anyway, if these logic is useful and we want to keep it as a building block (yet unused in high-level verification), it needs to be somehow tested.

File: src/verifier/step4/SumcheckLogic.sol [hunk 9]
```diff
@@ -0,0 +1,488 @@
...
+
+        bytes32 input_bytes = bytes32(input);
+
+        for (uint256 i = 0; i < 32; i++) {
+            result[i] = uint8(input_bytes[31 - i]);
+        }
+        return result;
+    }
+
+    function toTranscriptBytes(UniPoly memory poly) public pure returns (uint8[] memory) {
```
[the remark this answers is not in the document]
Comment (mpenciak) [reply src/verifier/step4/SumcheckLogic.sol:182]: I think that might introduce a dependency cycle `KeccakTranscript` -> `SumcheckLogic` -> `KeccakTranscript`.

A way around it is to separate out the polynomial library and sumcheck library files, which actually makes some sense. I'll try to do this refactor now.

File: src/verifier/step4/SumcheckLogic.sol [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+pragma solidity ^0.8.16;
+
+import "src/pasta/Pallas.sol";
+import "src/pasta/Vesta.sol";
+import "src/verifier/step4/KeccakTranscript.sol";
+import "src/verifier/step4/EqPolynomial.sol";
+import "src/Polynomial.sol";
+
+library PrimarySumcheck {
+    struct SumcheckProof {
```
Comment (storojs72) [inline src/verifier/step4/SumcheckLogic.sol:11]: I think this struct can also be moved to PolyLib as soon as it deals just with uint256s internally.

**Review by storojs72 — APPROVED** [review]
LGTM now! Thanks!

**Pull request merged**